pkg/ring/model.go

```diff
@@ -168,6 +191,39 @@ func (d *Desc) Ready(now time.Time, heartbeatTimeout time.Duration) error {
 	return nil
 }
 
+// SetIngesterTokens updates the tokens in the ring with the provided tokens. SetIngesterTokens
```

Comment:
So I have an idea ... all of these new methods that you're introducing to ring.Desc -- I understand why you need them but I also don't like that it bloats the ring code, structure that is stored in KV store, so much.

The fact that there is still `ring.Desc.Tokens` field is just a legacy, and we aim to get rid of it, as we don't want to store it in the KV. But obviously code working with the ring in memory operates on that.

I think we should split ring as stored in KV and ring as used in memory (basically `[]TokenDesc`) and have them as two separate types. Many of your new methods would go to the new type then, and we would be free to get rid of `ring.Desc.Tokens` eventually.